src/routing/error_cause.rs

```diff
@@ -129,13 +130,26 @@ impl ErrorCause {
             Self::RateLimited(_) => ErrorClientFacing::RateLimited,
             Self::HttpGatewayError(x) => match x {
                 HttpGatewayError::AgentError(y) => {
-                    let error_string = y.to_string();
-                    if error_string.contains("no_healthy_nodes") {
-                        return ErrorClientFacing::SubnetUnavailable;
-                    } else if error_string.contains("canister_not_found") {
-                        return ErrorClientFacing::CanisterIdNotFound;
+                    match y.as_ref() {
+                        AgentError::CertifiedReject(z) | AgentError::UncertifiedReject(z) => {
+                            if z.reject_code == RejectCode::CanisterError {
+                                return ErrorClientFacing::CanisterError("The canister encountered an error while processing the request.<br />This issue may be due to resource limitations, configuration problems, or an internal failure.".to_string())
+                            }
+                            else if z.reject_code == RejectCode::SysTransient && z.reject_message.contains("frozen") {
+                                return ErrorClientFacing::CanisterError("The canister is temporarily unable to process the request due to insufficient funds.".to_string())
+                            };
+                            ErrorClientFacing::UpstreamError
+                        },
+                        _ => {
+                            let error_string = y.to_string();
```

Comment:
i think one could avoid unnecessary overhead with conversion `to_string()`

Reply:
Changed it slightly, but could not completely get rid of the string matching due to how the error is.